paraatm/simulation_method/aviationr.py

```diff
@@ -0,0 +1,219 @@
+                if (runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                    break
+                if (runtime_sim_status == self.NATS_SIMULATION_STATUS_ENDED):
+                    print('Accident condition is not met!')
+                    return None
+            aclist = self.aircraftInterface.getAllAircraftId()
+            ac = self.aircraftInterface.select_aircraft(aclist[0])
+            phase = ac.getFlight_phase()
+            if accident_phase in list(ntsb2nats.loc[ntsb2nats['NATS_CODE'] == phase]['NTSB_CODE']): # test if the phase of flight from  NASTS simulation is same as NTSB
+                accident = True
+        # Start accident simulation and risk analysis
+        risk_list = [None] * len(case)
+        subject_list = [None] * len(case)
+        phase_list = [None] * len(case)
+        occurrence_list = [None] * len(case)
+        time_list = [None] * len(case)
+        risk_estimator = RiskEstimator(self.model, isRNN, self.data, device)
+        
+        model, hierarchical_softmax, risk_model = risk_estimator.load_model()
+        for i in range(1, len(case_code['Occurrence_Code'][0]) - 1):
+
+            # recording the risk and predicting events
+            risk_list[i - 1], phase_list[i - 1], occurrence_list[i - 1], subject_list[i - 1] = risk_estimator.risk_estimation(
+                case_code, i, model, hierarchical_softmax, risk_model, datadict, isRNN)
+
+            time_list[i-1] = int(self.simulationInterface.get_curr_sim_time()/30)
+            # modify NATS simulation according to NTSB accident
+            t = self.accident_simulator(self.aircraftInterface, self.controllerInterface, self.pilotInterface, phase_list[i-1],
+                                        occurrence_list[i-1], subject_list[i-1], datadict, ntsb2nats)
+            self.simulationInterface.resume(t)
+            while True:
+                runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+                if (runtime_sim_status == self.NATS_SIMULATION_STATUS_PAUSE):
+                    break
+                if (runtime_sim_status == self.NATS_SIMULATION_STATUS_ENDED):
+                    print('Simulation time not enough!')
+                    return None
+        self.simulationInterface.resume()
+        while True:
+            server_runtime_sim_status = self.simulationInterface.get_runtime_sim_status()
+            if (server_runtime_sim_status == self.NATS_SIMULATION_STATUS_ENDED):
+                break
+        event_list = []
+        for i in range(len(subject_list)):
+            event_list.append(phase_list[i] + '\n' + occurrence_list[i] + '\n' + subject_list[i])
+        print('Aviation accident simulation finished!')
+
+        if not output_dir:
+            output_dir = self.data+'result.csv'
+            
+        with open(output_dir, 'wb') as file:
+            pickle.dump({'risk':risk_list, 'event':event_list, 'time':time_list}, file, protocol=pickle.HIGHEST_PROTOCOL)
```

Comment:
Using file i/o is not a good way to return results from a function.  However, the problem really is that the `GnatsSimulationWrapper` API previously did not provide a way to propagate results from `simulation` to `__call__`.  I have updated the code so that it will allow you to do this.  You will need to first pull in the latest changes from the main repo:

If you have not already done so, add the main ("upstream") GitHub repository to your list of remotes:
```shell
git remote add upstream https://github.com/ymlasu/para-atm.git
```

Then, make sure that your working directory is clean (use `git status`) and pull in the latest changes:
```shell
git pull upstream master
```

Refer to [this guide](https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow) for more information on these steps.

With the latest changes in para-atm, `GnatsSimulationWrapper.__call__` now returns a dictionary with two keys: `'trajectory'` and `'sim_results'`.  The `'sim_results'` key will contain the return value from your `simulation` method.  You can change your `simulation` method to return the results as a dictionary, as your previously had.  Then, you can call your simulation as:
```python
results = sim(return_df=False)['sim_results']
```

This will invoke `__call__` and then extract the results of the `simulation` method from the dictionary entry in the `'sim_results'` key.  If you don't need the trajectory data, you can pass `return_df=False`, which prevents unnecessarily writing and reading the trajectory data.

See also the documentation [here](https://para-atm.readthedocs.io/en/latest/src/gnats.html#running-the-gnats-simulation), which has more detail regarding how to work with `GnatsSimulationWrapper.__call__`.

Reply:
Got it. Thanks for the modification. I have adjusted my code accordingly. Please let me know if it works.

Reply:
Thanks, it looks like you've got it set up correctly for returning the results.